simulator/task_manager.go

```diff
@@ -109,3 +179,220 @@ func (m *TaskManager) CreateTask(ctx *Context, req *types.CreateTask) soap.HasFa
 
 	return body
 }
+
+type TaskHistoryCollector struct {
+	mo.TaskHistoryCollector
+
+	*HistoryCollector
+}
+
+func (m *TaskManager) createCollector(ctx *Context, req *types.CreateCollectorForTasks) (*TaskHistoryCollector, *soap.Fault) {
+	if len(m.history.collectors) >= int(m.MaxCollector) {
+		return nil, Fault("Too many task collectors to create", new(types.InvalidState))
+	}
+
+	collector := &TaskHistoryCollector{
+		HistoryCollector: newHistoryCollector(ctx, m.history, defaultPageSize),
+	}
+	collector.Filter = req.Filter
+
+	return collector, nil
+}
+
+// taskFilterChildren returns true if a child of self is the task entity.
+func taskFilterChildren(ctx *Context, root types.ManagedObjectReference, task *types.TaskInfo) bool {
+	seen := false
+
+	var match func(types.ManagedObjectReference)
+
+	match = func(child types.ManagedObjectReference) {
+		if child == *task.Entity {
+			seen = true
+			return
+		}
+
+		walk(ctx.Map.Get(child), match)
+	}
+
+	walk(ctx.Map.Get(root), match)
+
+	return seen
+}
+
+// entityMatches returns true if the spec Entity filter matches the task entity.
+func (c *TaskHistoryCollector) entityMatches(ctx *Context, task *types.TaskInfo, spec *types.TaskFilterSpec) bool {
+	e := spec.Entity
+	if e == nil {
+		return true
+	}
+
+	isSelf := *task.Entity == e.Entity
+
+	switch e.Recursion {
+	case types.TaskFilterSpecRecursionOptionSelf:
+		return isSelf
+	case types.TaskFilterSpecRecursionOptionChildren:
+		return taskFilterChildren(ctx, e.Entity, task)
+	case types.TaskFilterSpecRecursionOptionAll:
+		return isSelf || taskFilterChildren(ctx, e.Entity, task)
+	}
+
+	return false
+}
+
+func (c *TaskHistoryCollector) stateMatches(_ *Context, task *types.TaskInfo, spec *types.TaskFilterSpec) bool {
+	if len(spec.State) == 0 {
```

Comment:
Is this desired? If there are no specified specs, then match everything?

Reply:
Yes, right. When `spec.State` _is_ specified, you can limit to tasks of state `failed` for example.